lib/ui/compositing.dart

```diff
@@ -79,16 +79,19 @@ class SceneBuilder extends NativeFieldWrapperClass2 {
   ///
   /// Rasterization outside the given rounded rectangle is discarded.
   ///
-  /// See [pop] for details about the operation stack.
-  void pushClipRRect(RRect rrect) => _pushClipRRect(rrect._value);
-  void _pushClipRRect(Float32List rrect) native 'SceneBuilder_pushClipRRect';
+  /// See [pop] for details about the operation stack, and [Clip] for different clip modes.
+  /// By default, the clip will be anti-aliased (clip = [Clip.antiAlias]).
+  void pushClipRRect(RRect rrect, {Clip clip = Clip.antiAlias}) => _pushClipRRect(rrect._value, clip.index);
+  void _pushClipRRect(Float32List rrect, int clipMode) native 'SceneBuilder_pushClipRRect';
 
   /// Pushes a path clip operation onto the operation stack.
   ///
   /// Rasterization outside the given path is discarded.
   ///
-  /// See [pop] for details about the operation stack.
-  void pushClipPath(Path path) native 'SceneBuilder_pushClipPath';
+  /// See [pop] for details about the operation stack. See [Clip] for different clip modes.
+  /// By default, the clip will be anti-aliased (clip = [Clip.antiAlias]).
+  void pushClipPath(Path path, {Clip clip = Clip.antiAlias}) => _pushClipPath(path, clip.index);
```

Comment:
what happens if you push a clip path with Clip.none?

Reply:
and does Clip.antiAliasWithSaveLayer do anything interesting here?

Reply:
Push a clip layer (e.g., ClipPath, ClipRRect) with Clip.none will trigger assert: https://github.com/flutter/engine/pull/5647/files#diff-9930584db1fb28d6064ad30841935ac1R51

Clip.antiAliasWithSaveLayer will add a Canvas.saveLayer call: https://github.com/flutter/engine/pull/5647/files#diff-9930584db1fb28d6064ad30841935ac1R55

Reply:
when you say "assert" do you mean "throw an AssertionError" or do you mean "crash the engine"?

Reply:
It's a `FXL_DCHECK(clip_mode_ != ClipMode::none)` just like `FXL_DCHECK(needs_painting())`. I think that it'll crash the engine.

Reply:
that seems very drastic. Normally we just assert in dart, that way people can hot-reload their way past the error.

Reply:
I'm confused. What stops someone from calling `SceneBuilder.pushClipPath(path, clip = Clip.none)` ?

Reply:
The ClipPathLayer in Dart (https://github.com/flutter/flutter/pull/18576/files#diff-58600d3d2645a66835eb03f3c41b447bR725) prevents C++ engine to have Clip.none ClipPath.

But now I see that if someone is pushing a ClipPathLayer without using the Dart class ClipPathLayer, then it could be a problem. I'll print a warning and drop the layer instead of crashing then.

Reply:
I would just assert here, just like we do in other methods in this file.

```dart
assert(clip != null);
assert(clip != Clip.none);
```

Reply:
Done.